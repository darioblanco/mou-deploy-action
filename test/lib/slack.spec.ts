import { oneLine, oneLineTrim } from 'common-tags';
import { context } from '@actions/github';
import { WebClient } from '@slack/web-api';

import {
  notifyDeploymentSuccess, notifyDeploymentError,
} from '@minddocdev/mou-deploy-action/lib/slack';

jest.mock('@actions/core');
jest.mock('@actions/github', () => ({
  context: {
    repo: { owner: 'myOwner', repo: 'myRepo' },
  },
}));
jest.mock('@slack/web-api');

describe('slack', () => {
  const repositoryPayload: any = {
    html_url: 'https://github.com/Codertocat/Hello-World',
  };
  const commitData = {
    author: {
      login: 'octocat',
      avatar_url: 'https://github.com/images/error/octocat_happy.gif',
      html_url: 'https://github.com/octocat',
    },
    commit: {
      message: 'Fix all the bugs',
      author: {
        name: 'Monalisa Octoca',
      },
    },
    html_url:
      'https://github.com/octocat/Hello-World/commit/6dcb09b5b57875f334f61aebed695e2e4193db5e',
  };
  const postMessage = jest.fn(() => ({ response_metadata: 'fake' }));
  (WebClient as any).mockImplementation(() => {
    return { chat: { postMessage } };
  });

  const getCommit = jest.fn(() => ({ data: commitData }));

  test('notify deployment success ', async () => {
    context.payload = {
      deployment: {
        sha: '6dcb09b5b57875f334f61aebed695e2e4193db5e',
        ref: 'master',
        id: 145988746,
      },
      repository: repositoryPayload,
    };
    const getReleaseByTag = jest.fn(() => ({
      data: {
        html_url: 'https://github.com/octocat/Hello-World/releases/v1.0.0',
        body: 'Description of the release',
      },
    }));
    const github = { repos: { getCommit, getReleaseByTag } };
    await notifyDeploymentSuccess(
      github as any,
      'production',
      'myapp',
      'http://url',
      { token: 'fakeToken', channel: 'fakeChannel' },
    );
    expect(postMessage.mock.calls[0]).toMatchSnapshot();
  });

  test('notify deployment error ', async () => {
    context.payload = {
      deployment: {
        sha: '6dcb09b5b57875f334f61aebed695e2e4193db5e',
        ref: '6dcb09b5b57875f334f61aebed695e2e4193db5e',
        id: 145988746,
      },
    };
    const getReleaseByTag = jest.fn(() => { throw new Error(); });
    const github = { repos: { getCommit, getReleaseByTag } };
    await notifyDeploymentError(
      github as any,
      {
        message: 'my fake error',
        stderr: 'stderr output',
        stdout: 'stdeout output',
        status: 1,
      },
      'staging',
      'myapp',
      {
        token: 'fakeToken',
        channel: 'fakeChannel',
      },
    );
    expect(postMessage.mock.calls[0]).toMatchSnapshot();
  });
});
