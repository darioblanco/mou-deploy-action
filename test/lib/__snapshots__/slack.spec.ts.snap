// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`slack notify deployment error  1`] = `
Array [
  Object {
    "blocks": Array [
      Object {
        "text": Object {
          "text": "I have FAILED to deploy *<https://github.com/myOwner/myRepo/commit/6dcb09b5b57875f334f61aebed695e2e4193db5e/checks|myapp>* (<https://github.com/myOwner/myRepo/commits/6dcb09b5b57875f334f61aebed695e2e4193db5e|6dcb09b...>) to *<https://github.com/myOwner/myRepo/deployments?environment=staging#activity-log|staging>* :shipit:",
          "type": "mrkdwn",
        },
        "type": "section",
      },
      Object {
        "type": "divider",
      },
      Object {
        "text": Object {
          "emoji": true,
          "text": "\`\`\`
Fix all the bugs
\`\`\`",
          "type": "plain_text",
        },
        "type": "section",
      },
      Object {
        "type": "divider",
      },
      Object {
        "accessory": Object {
          "alt_text": "avatar_url",
          "image_url": "https://github.com/images/error/octocat_happy.gif",
          "type": "image",
        },
        "fields": Array [
          Object {
            "text": ":books: *Repository:* <https://github.com/myOwner/myRepo|myRepo>",
            "type": "mrkdwn",
          },
          Object {
            "text": ":person_in_steamy_room: *Author:* Monalisa Octoca <https://github.com/octocat|@octocat>",
            "type": "mrkdwn",
          },
          Object {
            "text": ":beer: *Commit:* <https://github.com/octocat/Hello-World/commit/6dcb09b5b57875f334f61aebed695e2e4193db5e|6dcb09b>",
            "type": "mrkdwn",
          },
          Object {
            "text": ":point_left: *Deployment ID:* 145988746",
            "type": "mrkdwn",
          },
        ],
        "type": "section",
      },
      Object {
        "type": "divider",
      },
      Object {
        "text": Object {
          "emoji": true,
          "text": "my fake error (error code 1) \`stderr\`: \`\`\` stderr output \`\`\` \`stdout\`: \`\`\` stdeout output \`\`\`",
          "type": "plain_text",
        },
        "type": "section",
      },
      Object {
        "type": "divider",
      },
    ],
    "channel": "fakeChannel",
    "text": "Error deploying myapp to staging",
    "token": "fakeToken",
  },
]
`;

exports[`slack notify deployment success  1`] = `
Array [
  Object {
    "blocks": Array [
      Object {
        "text": Object {
          "text": "I have deployed *<https://github.com/Codertocat/Hello-World/commit/6dcb09b5b57875f334f61aebed695e2e4193db5e/checks|myapp>* (<https://github.com/octocat/Hello-World/releases/v1.0.0|master>) to *<https://github.com/Codertocat/Hello-World/deployments?environment=production#activity-log|production>* :rocket:",
          "type": "mrkdwn",
        },
        "type": "section",
      },
      Object {
        "type": "divider",
      },
      Object {
        "text": Object {
          "emoji": true,
          "text": "Description of the release",
          "type": "plain_text",
        },
        "type": "section",
      },
      Object {
        "type": "divider",
      },
      Object {
        "accessory": Object {
          "alt_text": "avatar_url",
          "image_url": "https://github.com/images/error/octocat_happy.gif",
          "type": "image",
        },
        "fields": Array [
          Object {
            "text": ":books: *Repository:* <https://github.com/Codertocat/Hello-World|myRepo>",
            "type": "mrkdwn",
          },
          Object {
            "text": ":person_in_steamy_room: *Author:* Monalisa Octoca <https://github.com/octocat|@octocat>",
            "type": "mrkdwn",
          },
          Object {
            "text": ":beer: *Commit:* <https://github.com/octocat/Hello-World/commit/6dcb09b5b57875f334f61aebed695e2e4193db5e|6dcb09b>",
            "type": "mrkdwn",
          },
          Object {
            "text": ":point_left: *Deployment ID:* 145988746",
            "type": "mrkdwn",
          },
        ],
        "type": "section",
      },
      Object {
        "elements": Array [
          Object {
            "text": "*Application url:* <http://url|url>",
            "type": "mrkdwn",
          },
        ],
        "type": "context",
      },
      Object {
        "type": "divider",
      },
    ],
    "channel": "fakeChannel",
    "text": "Deployed myapp to production",
    "token": "fakeToken",
  },
]
`;
